worlds/yachtdice/__init__.py

```diff
@@ -0,0 +1,459 @@
+            self.options.weight_of_fixed_score_multiplier.value,
+            self.options.weight_of_step_score_multiplier.value,
+            self.options.weight_of_double_category.value,
+            self.options.weight_of_points.value,
+        ]
+
+        # if the player wants extra rolls or dice, fill the pool with fragments until close to an extra roll/dice
+        if weights[0] > 0 and frags_per_dice > 1:
+            self.itempool += ["Dice Fragment"] * (frags_per_dice - 1)
+        if weights[1] > 0 and frags_per_roll > 1:
+            self.itempool += ["Roll Fragment"] * (frags_per_roll - 1)
+
+        # calibrate the weights, since the impact of each of the items is different
+        weights[0] = weights[0] / 5 * frags_per_dice
+        weights[1] = weights[1] / 5 * frags_per_roll
+
+        extra_points_added = 0
+        multipliers_added = 0
+        items_added = 0
+
+        def get_item_to_add():
+            nonlocal weights
+            nonlocal extra_points_added
+            nonlocal multipliers_added
+            nonlocal items_added
+
+            items_added += 1
+
+            all_items = self.itempool + self.precollected
+            dice_fragments_in_pool = all_items.count("Dice") * frags_per_dice + all_items.count("Dice Fragment")
+            if dice_fragments_in_pool + 1 >= 9 * frags_per_dice:
+                weights[0] = 0  # don't allow >=9 dice
+            roll_fragments_in_pool = all_items.count("Roll") * frags_per_roll + all_items.count("Roll Fragment")
+            if roll_fragments_in_pool + 1 >= 6 * frags_per_roll:
+                weights[1] = 0  # don't allow >= 6 rolls
+
+            # Don't allow too many multipliers
+            if multipliers_added > 50:
+                weights[2] = 0
+                weights[3] = 0
+
+            # Don't allow too many extra points
+            if extra_points_added > 300:
+                weights[5] = 0
+
+            # if all weights are zero, allow to add fixed score multiplier, double category, points.
+            if sum(weights) == 0:
+                if multipliers_added <= 50:
+                    weights[2] = 1
+                weights[4] = 1
+                if extra_points_added <= 300:
+                    weights[5] = 1
+
+            # Next, add the appropriate item. We'll slightly alter weights to avoid too many of the same item
+            which_item_to_add = self.multiworld.random.choices([0, 1, 2, 3, 4, 5], weights=weights)[0]
+            item_to_add = ""
+            if which_item_to_add == 0:
+                weights[0] /= 1 + frags_per_dice
+                if frags_per_dice == 1:
+                    return "Dice"
+                else:
+                    return "Dice Fragment"
+            elif which_item_to_add == 1:
+                weights[1] /= 1 + frags_per_roll
+                if frags_per_roll == 1:
+                    return "Roll"
+                else:
+                    return "Roll Fragment"
```

Comment:
I think this would look slightly nicer with a ternary?

`return "Roll" if frags_per_roll == 1 else "Roll Fragment"`

Personal preference tho

*Low Priority*